### ISSUE TITLE
USB HID

### ISSUE THREAD

**Issue by tannewt** [adafruit/circuitpython#1009]
https://circuitpython.readthedocs.io/en/latest/shared-bindings/usb_hid/__init__.html#module-usb_hid

This is wired USB HID.

**Comment by dhalbert** [comment]
split this into Bluetooth and wired USB support?

**Comment by tannewt** [comment]
This is USB only. A BLE HID issue would be good too.

**Comment by hathach** [comment]
Descriptors to support are
- keyboard
- mouse
- consumer key ( for volume up, down, brightness +/-)

**Comment by dhalbert** [comment]
We have code to generate the USB descriptors in https://github.com/adafruit/circuitpython/blob/master/ports/atmel-samd/tools/gen_usb_descriptor.py (most of which should maybe be moved out of the port-specific directory). For HID, that code calls the library code in https://github.com/adafruit/usb_descriptor(which is a submodule further up in the tree), including pre-made HID descriptors in https://github.com/adafruit/usb_descriptor/blob/master/adafruit_usb_descriptor/hid.py. Besides the ones you listed, we also support a generic gamepad. There is latent support for a touchscreen but it does not work on all OS platforms so I turned it off.

**Comment by hathach** [comment]
Ah thanks, I am also quite familiar with hid descriptor, so it is not really an issue. Ah, almost forgot the gamepad. I will add it as well

**Comment by dhalbert** [comment]
I figured you were an expert on that :grinning:. Let me know if you think there's something that could be improved in the current descriptors.

**Comment by hathach** [comment]
It is not true, I just copy thing here and there and put them together :)) . The whole hid descriptor is too complicated for anyone to digest :) . I will try out the descriptor generator as well. Above descriptors are all that I worked with so far.

**Comment by dhalbert** [comment]
If you want, you can use the descriptors in https://github.com/adafruit/usb_descriptor/blob/master/adafruit_usb_descriptor/hid.py. They are all lumped in a single-endpoint composite HID device. I spent a while tuning them up and making sure they work on all platforms.

**Comment by hathach** [comment]
@tannewt  no problem, I will make the hid report descriptor changeable dynamically (you will need to update the configuration descriptor for new length as well). However, to reduce amount of work, and get this done quickly, we will still implement it with fixed descriptor at first. When needed, we will change it with a separated PR.

**Comment by hathach** [comment]
@tannewt just out of curiousity, we want to change hid report dynamically is to switch on/offf keyboard, mouse, gamepad etc right ? I ask since I always want to have a high level of api as much as possible for tinyusb. Those can be achieved at some level with predefined hid descriptor, though the full hid parser is too much for a device to support.

PS for configuration descriptor it is already dynamic supported by tinyusb, it will reconfigure endpoints based on parsing config desc when enumeration :)

**Comment by hathach** [comment]
for ref
- https://github.com/DIGImend/usbhid-dump
- https://github.com/DIGImend/hidrd

**Comment by hathach** [comment]
@tannewt I see, class dynamic enable/disable is already support by tinyusb. python only need to feed the correct config descriptor. So it  is not a problem at all.

For HID, to construct report in Python, I feel like we don't need to dynamically change the report descriptor. I could be wrong, let's me get something running first to see what it turns out to be.

**Comment by dhalbert** [comment]
For HID, we mean adding or dropping devices from the multiple report descriptor: https://github.com/adafruit/usb_descriptor/blob/master/adafruit_usb_descriptor/hid.py#L170. E.g. turn off keyboard, or allow a user to add another device with its own Report ID. As you mention, this affects the length field in the composite descriptor. (This stuff was clearly designed by two different committees: it would be a lot easier if the length were in the header of the HID Report Descriptor :frowning_face:)

**Comment by hathach** [comment]
@dhalbert  thanks, I already got the idea, the C level should only need to implement a generic HID where user need to supply report descriptor and report, which I will added to the stack.

FYI, In my stack, originally it attempts to support higher API with both keyboard & mouse API. It is useful to user that is not familiar with  HID and only want to use out of the box keyboard/mouse.
https://github.com/hathach/tinyusb/blob/develop/src/class/hid/hid_device.h#L65

**Comment by hathach** [comment]
I also try to support boot keyboard at stack C-level as well, so it is a bit confusing with option and things. Trying to nail it now :D

**Comment by dhalbert** [comment]
Yeah, I don't think we need boot keyboard support, but it could be moved out of the multi-device HID descriptor to its own endpoint. We have enough endpoints on the SAMD, at least.

**Comment by dhalbert** [comment]
Note that you will not see this echoed in the REPL while it's running (if you just do `>>> k.send(Keycode.A)` you'll see it). But if you move the focus to another window you'll see 'a's being sent.
```
from adafruit_hid.keyboard import Keyboard
from adafruit_hid.keycode import Keycode
import time
k = Keyboard()
while True:
    k.send(Keycode.A)
    time.sleep(1.0)
```

**Comment by hathach** [comment]
Superb, thanks. Sorry for silly question, I haven't used cp much :(

**Comment by hathach** [comment]
:+1: :+1:

**Comment by hathach** [comment]
@dhalbert  the readthedoc link should list `Adafruit_CircuitPython_HID` as dependency instead of the core repo right ?

**Comment by hathach** [comment]
never mind, I think you mean the dependency of the HID library. Sorry, misunderstand there.

**Comment by hathach** [comment]
@dhalbert @tannewt I would like to light up an LED when receiving SET REPORT request from host e.g CAPSLOCK. But couldn't find where to add it, could you please give me a hint ?

**Comment by hathach** [comment]
keyboard and mouse works quite well, cleaning it up for full list of devices support

**Comment by dhalbert** [comment]
> @dhalbert @tannewt I would like to light up an LED when receiving SET REPORT request from host e.g CAPSLOCK. But couldn't find where to add it, could you please give me a hint ?

We haven't provided a way to do that reverse-direction dataflow at all yet in the HID module, so it would be noticeable work and an API addition. Fine to skip it for now. You could open a long-term issue

**Comment by hathach** [comment]
Ah, thanks. Since I don't know how to organize the cp API and my obj things. I will just turn on/off an LED0 for that, does it sound OK, if not I will leave it as stub.

**Comment by dhalbert** [comment]
A stub is fine too, since we don't want to reserve an LED just for this. It's more likely LED state would be sent on to some peripheral, (e.g. a pin) for someone who's building a keyboard. Right now CircuitPython doesn't have an of handling callbacks which we might want for this case (since SET REPORT can arrive spontaneously).

**Comment by hathach** [comment]
@dhalbert everything seems works, except for the gamepad, I got this error. I have no idea about this, am I missing something 

```
Adafruit CircuitPython 4.0.0-alpha-40-g2c85f42-dirty on 2018-07-31; PCA10056 with nRF52840                                                                                                                                                                                 
                                                                                                                                                                                                                                                                           
>>> from adafruit_hid.gamepad import Gamepad                                                                                                                                                                                                                               
                                                                                                                                                                                                                                                                           
>>> gp = Gamepad()                                                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                                           
Traceback (most recent call last):                                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                                           
  File "<stdin>", line 1, in <module>                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                           
  File "adafruit_hid/gamepad.py", line 89, in __init__                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                           
  File "adafruit_hid/gamepad.py", line 86, in __init__                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                           
  File "adafruit_hid/gamepad.py", line 149, in reset_all                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                           
  File "adafruit_hid/gamepad.py", line 163, in _send                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                           
TypeError: 'bytearray' object does not support item assignment                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                           
```

**Comment by dhalbert** [comment]
>TypeError: 'bytearray' object does not support item assignment

In `mpconfigport.h`, try changing these from `0` to `1`:
```
#define MICROPY_PY_ARRAY_SLICE_ASSIGN            (0)
...
#define MICROPY_PY_BUILTINS_SLICE_ATTRS (0)
```
There are probably some other changes to the feature on/off flags that I should make match up with atmel-samd and esp8266. I'll do some diff'ing.

**Comment by hathach** [comment]
Superb !! It is all working now, thanks. I will do some clean up and submit an PR for review

**Comment by tannewt** [comment]
@hathach Whatever works. :-) Feel free to change it.

**Comment by dhalbert** [comment]
Fixed by #1074.
